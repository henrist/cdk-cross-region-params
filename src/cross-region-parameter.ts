import * as cdk from "@aws-cdk/core"
import * as cr from "@aws-cdk/custom-resources"

export interface CrossRegionParameterProps {
  region: string
  name: string
  value: string
}

/**
 * SSM Parameter stored in another region.
 */
export class CrossRegionParameter extends cdk.Construct {
  constructor(
    scope: cdk.Construct,
    id: string,
    props: CrossRegionParameterProps,
  ) {
    super(scope, id)

    const physicalResourceId = cr.PhysicalResourceId.of(props.name)

    // TODO: Can we somehow propagate tags as well?
    new cr.AwsCustomResource(this, "Resoure", {
      onUpdate: {
        service: "SSM",
        action: "putParameter",
        parameters: {
          Name: props.name,
          Value: props.value,
          Type: "String",
          Overwrite: true,
        },
        region: props.region,
        physicalResourceId,
      },
      onDelete: {
        service: "SSM",
        action: "deleteParameter",
        parameters: {
          Name: props.name,
        },
        region: props.region,
        physicalResourceId,
      },
      policy: cr.AwsCustomResourcePolicy.fromSdkCalls({
        // We grant the Custom Resource access to write to any
        // parameter, as this is needed to be able to rename
        // a parameter later.
        resources: cr.AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
      installLatestAwsSdk: false,
    })
  }
}
