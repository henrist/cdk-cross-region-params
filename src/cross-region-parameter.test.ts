import { App, Stack } from "aws-cdk-lib"
import "jest-cdk-snapshot"
import { CrossRegionParameter } from "./cross-region-parameter"

test("cross-region-ssm-parameter", () => {
  const app = new App()
  const stack1 = new Stack(app, "Stack1", {
    env: {
      region: "us-east-1",
    },
  })

  new CrossRegionParameter(stack1, "Param", {
    name: "/param/name",
    region: "eu-west-1",
    value: "some value",
  })

  expect(stack1).toMatchCdkSnapshot({
    ignoreAssets: true,
  })
})
