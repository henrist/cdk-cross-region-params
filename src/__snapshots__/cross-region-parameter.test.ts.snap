// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cross-region-ssm-parameter 1`] = `
Object {
  "Parameters": Any<Object>,
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ParamResoure2E03152F": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "ParamResoureCustomResourcePolicy94B01452",
      ],
      "Properties": Object {
        "Create": "{\\"service\\":\\"SSM\\",\\"action\\":\\"putParameter\\",\\"parameters\\":{\\"Name\\":\\"/param/name\\",\\"Value\\":\\"some value\\",\\"Type\\":\\"String\\",\\"Overwrite\\":true},\\"region\\":\\"eu-west-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/param/name\\"}}",
        "Delete": "{\\"service\\":\\"SSM\\",\\"action\\":\\"deleteParameter\\",\\"parameters\\":{\\"Name\\":\\"/param/name\\"},\\"region\\":\\"eu-west-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/param/name\\"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{\\"service\\":\\"SSM\\",\\"action\\":\\"putParameter\\",\\"parameters\\":{\\"Name\\":\\"/param/name\\",\\"Value\\":\\"some value\\",\\"Type\\":\\"String\\",\\"Overwrite\\":true},\\"region\\":\\"eu-west-1\\",\\"physicalResourceId\\":{\\"id\\":\\"/param/name\\"}}",
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "ParamResoureCustomResourcePolicy94B01452": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:PutParameter",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "ssm:DeleteParameter",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ParamResoureCustomResourcePolicy94B01452",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
